File: strategies/SPYEMAStrategy.cs
using System;
using System.Collections.Generic;
using StrategyFramework.Builder;
using StrategyFramework.Interfaces;
using StrategyFramework.Metrics;
using StrategyFramework.Modules.Entries;
using StrategyFramework.Modules.Exits;
using StrategyFramework.Modules.Sizing;
using StrategyFramework.Modules.Risk;

namespace StrategyFramework.Strategies
{
    /// <summary>
    /// SPY EMA Crossover Strategy Test
    /// Entry: Price > EMA50
    /// Exit: Price < EMA100
    /// Allocation: 100% of portfolio
    /// Timeframe: Daily
    /// Period: Q3 2025
    /// Starting Capital: $1M
    /// </summary>
    public static class SPYEMAStrategy
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("SPY EMA CROSSOVER STRATEGY - Q3 2025 BACKTEST");
            Console.WriteLine("================================================");
            Console.WriteLine("Strategy Rules:");
            Console.WriteLine("   * BUY SPY when close > EMA50");
            Console.WriteLine("   * SELL SPY when close < EMA100");
            Console.WriteLine("   * Allocation: 100% of portfolio");
            Console.WriteLine("   * Timeframe: Daily");
            Console.WriteLine("   * Period: Q3 2025 (Jul-Sep)");
            Console.WriteLine("   * Starting Capital: $1,000,000");
            Console.WriteLine();

            // Build the strategy
            var entryModule = new EMA50Entry();
            var exitModule = new EMA100Exit();
            
            var strategy = new SimpleStrategyBuilder()
                .WithName("SPY EMA50/100 Crossover")
                .WithEntry(entryModule)
                .WithExit(exitModule)
                .WithPositionSizing(new FullAllocationSizing())
                .WithRiskManagement(new BasicRiskManagement())
                .Build();

            Console.WriteLine($"Strategy Built: {strategy.Name}");
            Console.WriteLine($"   Entry: {strategy.EntryModule.Name}");
            Console.WriteLine($"   Exit: {strategy.ExitModule.Name}");
            Console.WriteLine($"   Sizing: {strategy.PositionSizingModule.Name}");
            Console.WriteLine($"   Risk: {strategy.RiskModule.Name}");
            Console.WriteLine();

            // Setup simulation
            var startingCapital = 1_000_000m;
            var results = new QuickResults(startingCapital);
            var positions = new Dictionary<string, decimal>();
            var portfolioValue = startingCapital;
            var availableCash = startingCapital;
            
            // Track entry details for exit calculation
            var entryPrice = 0m;
            var entryTime = DateTime.MinValue;

            // Simulate Q3 2025 daily data (realistic SPY price action)
            var q3Data = GenerateQ3_2025_SPYData();

            Console.WriteLine("Running Backtest...");
            Console.WriteLine("Date        | Close   | EMA50   | EMA100  | Action");
            Console.WriteLine("------------|---------|---------|---------|------------------");

            var q3Start = new DateTime(2025, 7, 1);
            
            foreach (var data in q3Data)
            {
                var currentPosition = positions.GetValueOrDefault("SPY", 0);
                
                // Update portfolio value
                if (currentPosition != 0)
                {
                    portfolioValue = availableCash + (currentPosition * data.Close);
                }

                // Process strategy
                var orders = strategy.ProcessData(data, positions, portfolioValue, availableCash);

                var action = "HOLD";
                foreach (var order in orders)
                {
                    if (order.Tag.StartsWith("Entry"))
                    {
                        // Entry trade
                        positions["SPY"] = order.Quantity;
                        availableCash -= order.Quantity * data.Close;
                        entryPrice = data.Close;
                        entryTime = data.Timestamp;
                        action = $"BUY {order.Quantity:F0} shares";
                    }
                    else if (order.Tag.StartsWith("Exit"))
                    {
                        // Exit trade
                        availableCash += currentPosition * data.Close;
                        
                        // Record completed trade
                        results.AddTrade("SPY", entryTime, data.Timestamp, entryPrice, data.Close, currentPosition);
                        
                        positions["SPY"] = 0;
                        action = $"SELL {currentPosition:F0} shares";
                    }
                }

                // Only show Q3 data in output (but process all data for EMA calculation)
                if (data.Timestamp >= q3Start)
                {
                    // Get EMA values for display
                    var ema50 = entryModule.GetCurrentEMA();
                    var ema100 = exitModule.GetCurrentEMA();

                    Console.WriteLine($"{data.Timestamp:MM/dd/yyyy} | ${data.Close:F2}  | ${ema50:F2}  | ${ema100:F2}  | {action}");
                }
            }

            // Final portfolio value
            var finalPosition = positions.GetValueOrDefault("SPY", 0);
            if (finalPosition != 0)
            {
                var finalData = q3Data[q3Data.Count - 1];
                portfolioValue = availableCash + (finalPosition * finalData.Close);
            }
            else
            {
                portfolioValue = availableCash;
            }

            Console.WriteLine();
            Console.WriteLine("BACKTEST RESULTS:");
            Console.WriteLine("===================");
            Console.WriteLine($"Starting Capital: ${startingCapital:N0}");
            Console.WriteLine($"Ending Portfolio: ${portfolioValue:N0}");
            Console.WriteLine($"Total Return: {((portfolioValue - startingCapital) / startingCapital):P2}");
            Console.WriteLine();

            // Show detailed metrics if we had trades
            if (results.TotalTrades > 0)
            {
                Console.WriteLine(results.GetSummary());
            }
            else
            {
                Console.WriteLine("No completed trades during the period.");
                Console.WriteLine("(This could mean the strategy stayed in a position through Q3)");
            }

            Console.WriteLine();
            Console.WriteLine("STRATEGY TEST COMPLETE!");
            Console.WriteLine("Ready to implement in QuantConnect for real backtesting!");
        }

        private static List<SecurityData> GenerateQ3_2025_SPYData()
        {
            var data = new List<SecurityData>();
            
            // Start earlier to warm up the EMAs (need 100+ days for EMA100)
            var startDate = new DateTime(2025, 3, 1); // Start in March for warm-up
            var endDate = new DateTime(2025, 9, 30);  // Q3 end
            
            var basePrice = 450m; // Realistic SPY price for 2025
            var currentDate = startDate;
            var price = basePrice;
            var random = new Random(42); // Fixed seed for reproducible results

            // Create a trending market with some volatility to generate crossovers
            var trendPhase = 0; // 0 = down, 1 = up, 2 = sideways
            var daysSincePhaseChange = 0;

            while (currentDate <= endDate)
            {
                // Skip weekends
                if (currentDate.DayOfWeek != DayOfWeek.Saturday && currentDate.DayOfWeek != DayOfWeek.Sunday)
                {
                    daysSincePhaseChange++;
                    
                    // Change trend phases to create crossovers
                    if (daysSincePhaseChange > 30)
                    {
                        trendPhase = (trendPhase + 1) % 3;
                        daysSincePhaseChange = 0;
                    }
                    
                    // Generate price movement based on trend phase
                    double trendComponent = 0;
                    switch (trendPhase)
                    {
                        case 0: // Down trend
                            trendComponent = -0.5; // Stronger downtrend
                            break;
                        case 1: // Up trend  
                            trendComponent = 0.6; // Stronger uptrend
                            break;
                        case 2: // Sideways
                            trendComponent = 0.0;
                            break;
                    }
                    
                    // Add random daily movement
                    var randomComponent = (random.NextDouble() * 0.04 - 0.02); // +/- 2% random
                    var dailyReturn = (decimal)(trendComponent / 100 + randomComponent);
                    
                    price = price * (1 + dailyReturn);
                    
                    // Keep price in reasonable range
                    price = Math.Max(350m, Math.Min(550m, price));

                    data.Add(new SecurityData
                    {
                        Symbol = "SPY",
                        Close = Math.Round(price, 2),
                        Open = Math.Round(price * 0.999m, 2),
                        High = Math.Round(price * 1.005m, 2),
                        Low = Math.Round(price * 0.995m, 2),
                        Volume = 50_000_000 + random.Next(0, 20_000_000),
                        Timestamp = currentDate
                    });
                }
                currentDate = currentDate.AddDays(1);
            }

            return data;
        }

        private static decimal GetSimpleEMA(decimal currentPrice, int period)
        {
            // Simplified EMA calculation for display purposes
            // In real implementation, this would use the actual EMA calculation from modules
            return currentPrice * 0.98m; // Approximate for demo
        }
    }
}

File: strategies/spy_ema_strategy.py
"""
SPY EMA Crossover Strategy

Strategy Rules:
- BUY SPY when close price crosses ABOVE EMA50
- SELL SPY when close price crosses BELOW EMA100  
- Position Size: 99% of available cash
- Timeframe: Daily

Entry: Price > EMA50
Exit: Price < EMA100
"""

from AlgorithmImports import *
from framework.interfaces import IStrategy


class SPYEMAStrategy(IStrategy):
    """SPY EMA Crossover Strategy Implementation"""
    
    def initialize(self, algorithm):
        """Initialize the SPY EMA strategy"""
        # Set algorithm framework settings
        algorithm.set_start_date(2024, 1, 1)   # Clean 2024 year
        algorithm.set_end_date(2024, 12, 31)   # Clean 2024 year  
        algorithm.set_cash(100_000)            # $100K starting capital
        
        # Add SPY with daily resolution
        self.spy = algorithm.add_equity("SPY", Resolution.DAILY).symbol
        
        # Create EMA indicators
        self.ema50 = algorithm.ema(self.spy, 50, Resolution.DAILY)
        self.ema100 = algorithm.ema(self.spy, 100, Resolution.DAILY)
        
        # Track our entry price for logging
        self.entry_price = None
        self.entry_time = None
        
        # Track SPY prices for buy-and-hold comparison
        self.spy_prices = []
        self.trading_days = []
        
        # Algorithm parameters
        self.allocation_percent = 0.99  # Use 99% to allow for fees
        
        # Warm up indicators (need 100 days for EMA100)
        algorithm.set_warm_up(105)  # Extra buffer for stability
        
        # Log strategy parameters
        algorithm.log(f" SPY EMA CROSSOVER STRATEGY INITIALIZED")
        algorithm.log(f"    Period: {algorithm.start_date} to {algorithm.end_date}")
        algorithm.log(f"    Starting Capital: ${algorithm.portfolio.cash:,.0f}")
        algorithm.log(f"    Entry: Price > EMA50, Exit: Price < EMA100")
        algorithm.log(f"    Allocation: {self.allocation_percent:.0%}")

    def on_data(self, algorithm, data):
        """Process market data for SPY EMA strategy"""
        # Skip if warming up or no data
        if algorithm.is_warming_up or not data.contains_key(self.spy):
            return
            
        # Skip if indicators not ready
        if not (self.ema50.is_ready and self.ema100.is_ready):
            return
            
        # Get current data - add safety check
        spy_data = data.get(self.spy)
        if spy_data is None:
            return
            
        current_price = spy_data.close
        
        # Track SPY prices for buy-and-hold calculation
        self.spy_prices.append(float(current_price))
        self.trading_days.append(algorithm.time)
        
        ema50_value = self.ema50.current.value
        ema100_value = self.ema100.current.value
        current_holdings = algorithm.portfolio[self.spy].quantity
        
        # ENTRY LOGIC: Buy when price crosses above EMA50 (and we're not already long)
        if current_holdings == 0:  # No position
            if current_price > ema50_value:
                # Calculate position size - use 99% of available cash
                available_cash = algorithm.portfolio.cash * self.allocation_percent
                quantity = int(available_cash / current_price)
                
                if quantity > 0:
                    algorithm.market_order(self.spy, quantity)
                    self.entry_price = current_price
                    self.entry_time = algorithm.time
                    
                    algorithm.log(f" BUY SIGNAL: {quantity:,} shares SPY @ ${current_price:.2f}")
                    algorithm.log(f"    EMA50: ${ema50_value:.2f} | EMA100: ${ema100_value:.2f}")
                    algorithm.log(f"    Position Value: ${quantity * current_price:,.0f}")
        
        # EXIT LOGIC: Sell when price crosses below EMA100 (and we have a position)
        elif current_holdings > 0:  # Have long position
            if current_price < ema100_value:
                # Close entire position
                algorithm.market_order(self.spy, -current_holdings)
                
                # Calculate trade performance
                if self.entry_price:
                    trade_return = (current_price - self.entry_price) / self.entry_price
                    trade_pnl = current_holdings * (current_price - self.entry_price)
                    hold_days = (algorithm.time - self.entry_time).days if self.entry_time else 0
                    
                    algorithm.log(f" SELL SIGNAL: {current_holdings:,} shares SPY @ ${current_price:.2f}")
                    algorithm.log(f"   EMA50: ${ema50_value:.2f} | EMA100: ${ema100_value:.2f}")
                    algorithm.log(f"   Trade Return: {trade_return:.2%} | P&L: ${trade_pnl:,.0f}")
                    algorithm.log(f"   Hold Period: {hold_days} days")
                    
                self.entry_price = None
                self.entry_time = None
        
        # Log daily status (only when we have a position or significant moves)
        if current_holdings > 0 or algorithm.time.day == 1:  # Monthly status updates
            portfolio_value = algorithm.portfolio.total_portfolio_value
            unrealized_pnl = 0
            if self.entry_price and current_holdings > 0:
                unrealized_pnl = current_holdings * (current_price - self.entry_price)
                
            algorithm.log(f" Status: Portfolio=${portfolio_value:,.0f} | SPY=${current_price:.2f} | "
                    f"EMA50={ema50_value:.2f} | EMA100={ema100_value:.2f} | "
                    f"Shares={current_holdings:,} | Unrealized=${unrealized_pnl:,.0f}")

    def on_end_of_algorithm(self, algorithm):
        """Called when the algorithm terminates - log final performance vs Buy & Hold"""
        final_value = algorithm.portfolio.total_portfolio_value
        strategy_return = (final_value - 1_000_000) / 1_000_000
        
        # Calculate Buy and Hold performance from price history
        if len(self.spy_prices) >= 2:
            start_price = self.spy_prices[0]
            end_price = self.spy_prices[-1]
            buy_hold_return = (end_price - start_price) / start_price
            
            # Calculate what the buy-and-hold value would be
            initial_investment = 1_000_000 * self.allocation_percent
            buy_hold_shares = initial_investment / start_price
            buy_hold_final_value = buy_hold_shares * end_price
            
            outperformance = strategy_return - buy_hold_return
            trading_days = len(self.spy_prices)
        else:
            buy_hold_return = 0
            outperformance = strategy_return
            buy_hold_final_value = 1_000_000
            start_price = end_price = 0
            trading_days = 0
        
        algorithm.log(f"\n BACKTEST COMPLETE - PERFORMANCE COMPARISON")
        algorithm.log(f"=" * 60)
        algorithm.log(f"    STRATEGY PERFORMANCE:")
        algorithm.log(f"      Starting Capital: $1,000,000")
        algorithm.log(f"      Final Portfolio Value: ${final_value:,.0f}")
        algorithm.log(f"      Total Return: {strategy_return:.2%}")
        algorithm.log(f"")
        algorithm.log(f"    BUY & HOLD BENCHMARK (SPY):")
        algorithm.log(f"      SPY Start Price: ${start_price:.2f}")
        algorithm.log(f"      SPY End Price: ${end_price:.2f}")
        algorithm.log(f"      Buy & Hold Return: {buy_hold_return:.2%}")
        algorithm.log(f"      Buy & Hold Final Value: ${buy_hold_final_value:,.0f}")
        algorithm.log(f"")
        algorithm.log(f"    COMPARISON:")
        if outperformance > 0:
            algorithm.log(f"      Strategy OUTPERFORMED by {outperformance:.2%}")
        elif outperformance < 0:
            algorithm.log(f"      Strategy UNDERPERFORMED by {abs(outperformance):.2%}")
        else:
            algorithm.log(f"      = Strategy matched Buy & Hold")
        algorithm.log(f"")
        algorithm.log(f"    TRADING STATISTICS:")
        algorithm.log(f"      Trading Days: {trading_days}")
        if trading_days > 0:
            annualized_strategy = ((final_value / 1_000_000) ** (252 / trading_days)) - 1
            annualized_buy_hold = ((1 + buy_hold_return) ** (252 / trading_days)) - 1
            algorithm.log(f"      Annualized Strategy Return: {annualized_strategy:.2%}")
            algorithm.log(f"      Annualized Buy & Hold Return: {annualized_buy_hold:.2%}")
        algorithm.log(f"=" * 60)
        
        # Final position status
        spy_holdings = algorithm.portfolio[self.spy].quantity
        if spy_holdings > 0:
            current_price = algorithm.securities[self.spy].close
            algorithm.log(f"    Final Position: {spy_holdings:,} shares @ ${current_price:.2f}")
        else:
            algorithm.log(f"    Final Position: CASH")


# For the active_strategy.py pattern, we also need to export the class as ActiveStrategy
class ActiveStrategy(SPYEMAStrategy):
    """Active strategy alias for SPY EMA Strategy"""
    pass